### PR TITLE
Changelog for 3.12.0 beta

### PR DESCRIPTION
**Pull request by Lex-2008**
and 3.11.0 which should be on master

### CODE REVIEW

**Comment by nickanderson** [comment]
lets squash these commits.

Waiting on anything special before merge?

**Comment by jimis** [comment]
Closing since alpha was not released externally. 3.12.0 Beta changelog should be in the 3.12.x branch.

File: ChangeLog [hunk 1]
```diff
@@ -1,3 +1,177 @@
+3.12.0b1:
+	New Features:
+	- Add an --key-type option to specify RSA key size to cf-key
```
Comment (olehermanse) [inline ChangeLog]: Replace `an` with `a` or put `--key-type` in parenthesis after `option`.

File: ChangeLog [hunk 2]
```diff
@@ -1,3 +1,177 @@
...
+	- Speedup evalution by not copying variables table when expanding a promise
+	  (CFE-2524)
+	- Resolve subkey conflicts when converting to JSON
+	  Whenever there is a conflict of array variable definitions prefer
+	  the container subkeys over simple values when converting to JSON.
+	  (CFE-2536)
+	- Do not ignore meta promises in server bundles (CFE-2066)
+	- Add a debug log for computed class in splayclass
+	- Dynamically grow HashMap on inserts (CFE-2301)
+	- Dynamically shrink HashMap when items are removed (CFE-2301)
```
Comment (olehermanse) [inline ChangeLog]: These HashMap changes aren't really user facing changes? Maybe update the message to reflect what this means for users, or remove them?

File: ChangeLog [hunk 3]
```diff
@@ -1,3 +1,177 @@
...
+	- Don't error when calling isexecutable on broken link (CFE-741)
+	- Open measurements promise type from enterprise cf-monitord.
```
Comment (olehermanse) [inline ChangeLog]: Looks weird that some lines end in dot, while others don't. Remove period here, as well as in other places.

File: ChangeLog (hunk 2 above)
Comment (Lex-2008) [reply ChangeLog]: I'm not sure what it means for the user, so will just delete these entries

File: ChangeLog [hunk 4]
```diff
@@ -1,3 +1,167 @@
...
+	  not occuring during signal handler cleanup. (CFE-1996)
+	- Fix a bug which could cause cf-execd to believe there was
+	  an error when sending the email report, when there really wasn't.
+	- zendesk#3204: Fix "lastseenexpireafter" 32-bit signed int overflow.
+	- fix cf-execd not exiting immediately with SIGTERM on AIX. (ENT-3147)
+	- Fix automatic service stops based on runlevel (redhat/centos)
+	  (CFE-2611)
+	- fix cf-serverd crash when reporting corrupted data. (ENT-3023)
+	- Fix rare output truncation on Solaris 10/11 (CFE-2527)
+	- Fix crash on Solaris when ps ucb variant is not available. (CFE-2506)
```
Comment (olehermanse) [inline ChangeLog]: Remove trailing dot.

File: ChangeLog [hunk 5]
```diff
@@ -1,3 +1,167 @@
...
+	- fix insert_lines related memory corruption (CFE-2520)
+	- prevent LMDB assertion on AIX by ensuring nested DB calls are
+	  not occuring during signal handler cleanup. (CFE-1996)
+	- Fix a bug which could cause cf-execd to believe there was
+	  an error when sending the email report, when there really wasn't.
+	- zendesk#3204: Fix "lastseenexpireafter" 32-bit signed int overflow.
+	- fix cf-execd not exiting immediately with SIGTERM on AIX. (ENT-3147)
+	- Fix automatic service stops based on runlevel (redhat/centos)
+	  (CFE-2611)
+	- fix cf-serverd crash when reporting corrupted data. (ENT-3023)
```
Comment (olehermanse) [inline ChangeLog]: Remove trailing dot and capitalize first letter.

File: ChangeLog [hunk 6]
```diff
@@ -1,3 +1,167 @@
...
+	- Fix rare output truncation on Solaris 10/11 (CFE-2527)
+	- Fix crash on Solaris when ps ucb variant is not available. (CFE-2506)
+	- Fix logic to detect when running under a Xen Hypervisor (CFE-1563)
+	- Fix "lastseenexpireafter" 32-bit signed int overflow. (zendesk#3204)
```
Comment (olehermanse) [inline ChangeLog]: Remove trailing dot.

File: ChangeLog [hunk 7]
```diff
@@ -1,3 +1,167 @@
...
+	- Print verbose instead of error message when readfile() fails.
+	  (CFE-2512)
+	- cf-serverd: Auto configure max open files ulimit according to
+	  maxconnections. (CFE-2575)
+	- Made the max bytes parameter to file reading functions optional.
+	  Affects readfile(), readenvfile(), readcsv()
+
+	Bug fixes:
+	- fix insert_lines related memory corruption (CFE-2520)
+	- prevent LMDB assertion on AIX by ensuring nested DB calls are
```
Comment (olehermanse) [inline ChangeLog]: Capitalize first letter.

File: ChangeLog [hunk 8]
```diff
@@ -1,3 +1,167 @@
...
+	  errors (CFE-2696)
+	- Fix rare cf-execd hang (CFE-2719)
+	- Fix mergedata segfault when called on a non-container (CFE-2704)
+	- Do not segfault if policy_server.dat only contains whitespaces and/or line breaks
+	- Fix segfault on JSON policy files with no bundles and bodies (CFE-2754)
+
+
+3.11.0:
+	New Features:
+	- allow function calls in promiser using universal "with" attribute.
```
Comment (olehermanse) [inline ChangeLog]: Capitalize first letter.

File: ChangeLog [hunk 9]
```diff
@@ -1,3 +1,167 @@
...
+	  the container subkeys over simple values when converting to JSON
+	  (CFE-2536)
+	- Do not ignore meta promises in server bundles (CFE-2066)
+	- Add a debug log for computed class in splayclass
+	- Don't error when calling isexecutable on broken link (CFE-741)
+	- Fix segfault when no show-evaluated-vars/classes is specified
+	- Fix memory leak in cf-execd, triggered when sending email failed (CFE-2712)
+	- fix IPv6 parsing to be un-reversed (CFE-2580)
+	- Fix bug preventing permission changes on Unix sockets (CFE-1782)
+	- fix storage mount promise when existing mountpoint has a similar path
```
Comment (olehermanse) [inline ChangeLog]: Capitalize first letter.

File: ChangeLog [hunk 10]
```diff
@@ -1,3 +1,167 @@
...
+	- Resolve subkey conflicts when converting to JSON
+	  Whenever there is a conflict of array variable definitions prefer
+	  the container subkeys over simple values when converting to JSON
+	  (CFE-2536)
+	- Do not ignore meta promises in server bundles (CFE-2066)
+	- Add a debug log for computed class in splayclass
+	- Don't error when calling isexecutable on broken link (CFE-741)
+	- Fix segfault when no show-evaluated-vars/classes is specified
+	- Fix memory leak in cf-execd, triggered when sending email failed (CFE-2712)
+	- fix IPv6 parsing to be un-reversed (CFE-2580)
```
Comment (olehermanse) [inline ChangeLog]: Capitalize first letter.

File: ChangeLog [hunk 11]
```diff
@@ -1,3 +1,167 @@
...
+	- New sysctlvalue() and data_sysctlvalues() functions from /proc/sys
+	  (CFE-2513)
+	- readdata() also auto-detects .yml files as YAML
+	- Added support for ENV and CSV file parsing (CFE-1881)
+	- Added vars and classes for CoreOS (ENT-3043)
+	- cf-agent: implement --show-evaluated-vars and --show-evaluated-classes
+	- Support for custom ports and host names as policy hub (CFE-953)
+	- cf-promises: allows --show-vars and --show-classes to take an optional filter
+	- Added a new tool: cf-net. cf-net is a CLI for the CFEngine
+	  network protocol, useful for debugging, testing etc.
```
Comment (olehermanse) [inline ChangeLog]: I would use this description of cf-net in the New Features section above about cf-net.

Comment (Lex-2008) [reply ChangeLog]: hm, but did cf-net appeared first in 3.11 or 3.12? According to 3.11 changelog, it was already there. Then want is it doing in 3.12 changelog? Or do we want to remove it from 3.11 changelog, in order to "promote" by putting to the top of changelog file?

Comment (olehermanse) [reply ChangeLog]: In your current changelog it appears under "New Features" in 3.12 and regular changelog of 3.11. I though this was to highlight it in 3.12, I just disliked the description there. I suggest using this description in both places.

Comment (Lex-2008) [reply ChangeLog]: ok!

**Pull request merged**